### PR TITLE
ReadHistory: On remove, trust the index from FileManagerHistory if possible

### PR DESCRIPTION
**Pull request by NiLuJe**
No need to look it up when we already provide it ;).

(Noticed because I had an item that refused to get removed on an old device. And because I'm an idiot, I fixed it without backing up the history.lua -_-".)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12253)

### CODE REVIEW

**Review by Frenzie — APPROVED** [review]

**Comment by hius07** [comment]
I'm afraid this doesn't work if we show the filtered history list.

**Comment by NiLuJe** [comment]
Would the index be missing (because that's fine), or *wrong* (because that's not ;p)?

**Comment by hius07** [comment]
If the filtered list is shown, Menu `item_table` is not the same as `readhistory.hist`.

**Comment by NiLuJe** [comment]
Oh, I hadn't seen your edit, just the original email notification, thanks ;).

It looks like checking whether filtering is at play is much simpler than your bookmark example ;).

**Comment by hius07** [comment]
No, we also can filter by collections and by a search string.

**Comment by hius07** [comment]
https://github.com/koreader/koreader/blob/9b9898d4665167740442cccf43f895401289b6b9/frontend/apps/filemanager/filemanagerhistory.lua#L97-L121

**Comment by NiLuJe** [comment]
And I have another file that `getIndexByTime` fails to find on that device, so I'll look into that, too ;).

**Comment by NiLuJe** [comment]
Oh, I know why it doesn't find it: I have out-of-order entries in that history file o_O.

(i.e., the timestamp of the affected books is much lower that what's on *either* side of them).

**Comment by NiLuJe** [comment]
Specifically, stuff with a timestamp around `Sun Jan  1 03:20:31 AM CET 2012` (1325384431), which, err, sounds odd ^^.

For reference, that's my Forma, which is now super-crashy, so it's entirely plausible I'd touched stuff after a crash with broken date/time set...

(The original file that tripped me up was KOReader's crash.log, but this is an actual book. Possibly the last one open'ed during a crash).

**Comment by poire-z** [comment]
@hius07 : just asking: how everything (history, highlight) survive the time going back and highlights made and books opened in that state, in regards to sorting happening or not, and/or new stuff being explicitely inserted at the head or the tail ?
It has happened to me twice that my battery went to zero or something, and after charging, I had the time set to 2018 (or some years ago), and it took me a few days to realize that :)

**Comment by NiLuJe** [comment]
That was exactly what I was going to ask ;p.

I vaguely remember that, at one point in time, flush did an explicit re-sort by time, right?

**Comment by poire-z** [comment]
There was some issue I noticed with .old being used instead of the newer, and that you fixed as part of #9902.

**Comment by hius07** [comment]
Annotations are sorted by position. Datetime doesn't matter.

History must be sorted by datetime, new entries are added on top. If the device time is broken, new entry is added with a wrong datetime and further binary searches will fail. No checks, repairing or resorting are done.

**Comment by hius07** [comment]
For History we can
(1) when adding a new entry, compare current time with the latest entry, and give a warning to draw the attention to the device time. This might require the list resorting afterwards.
(2) silently insert a new entry to the appropriate position according to the current (wrong) time. This would keep the list ordered.

**Comment by NiLuJe** [comment]
Both options end up with the history sorted properly: i.e., in our "battery dead, rtc reset" example, highly likely at the *end*, so I'm not necessarily sure we need an extra warning?

TL;DR: 2. sounds fine to me.

**Comment by NiLuJe** [comment]
I was going to say that 1. might actually be *cheaper* than 2. for 99.99% of the cases, as we don't need to compute the insertion index, just to compare now vs. hist[1].time, and keep using a simple table.insert() ... but then I remembered history is sorted in revesre order, so we actually do table.insert(1, ...), so the cost point is moot, as that already sucks for performance anyway ;p.

**Comment by poire-z** [comment]
Somehow, I prefer 1) and I like the warning :)
May be adding a key `ordering_possibly_messed_up = true` to the history array so we can keep showing the warning and propose the re-ordering via a ConfirmBox ?

**Comment by NiLuJe** [comment]
I'm fine with both approaches, so if others prefer 1., that's perfectly fine with me ;).

**Comment by Frenzie** [comment]
I suppose I would theoretically prefer 1 since something "bad" is happening but for this scenario I doubt I care in practice.

**Review by hius07 — APPROVED** [review]
